near/receiver/src/governance.rs

```diff
@@ -0,0 +1,385 @@
+
+        match self {
+            Self::ContractUpgrade(hash) => {
+                data.extend_from_slice(hash);
+            }
+
+            Self::SetDataSources(sources) => {
+                data.push(sources.len() as u8);
+                for source in sources {
+                    data.extend_from_slice(&source.emitter);
+                    data.extend_from_slice(&source.pyth_emitter_chain.0.to_le_bytes());
+                }
+            }
+
+            Self::SetGovernanceSource(source) => {
+                data.extend_from_slice(&source.emitter);
+                data.extend_from_slice(&source.pyth_emitter_chain.0.to_le_bytes());
+            }
+
+            Self::SetStalePriceThreshold(stale_price_threshold) => {
+                data.extend_from_slice(&stale_price_threshold.to_le_bytes());
+            }
+
+            Self::SetUpdateFee(update_fee) => {
+                data.extend_from_slice(&update_fee.to_le_bytes());
+            }
+        }
+
+        data
+    }
+}
+
+#[near_bindgen]
+impl Pyth {
+    /// Instruction for processing Governance VAA's relayed via Wormhole.
+    ///
+    /// Note that VAA verification requires calling Wormhole so processing of the VAA itself is
+    /// done in a callback handler, see `process_vaa_callback`.
+    #[payable]
+    #[handle_result]
+    pub fn execute_governance_instruction(&mut self, vaa: String) -> Result<(), Error> {
+        // Verify the VAA is coming from a trusted source chain before attempting to verify VAA
+        // signatures. Avoids a cross-contract call early.
+        {
+            let vaa = hex::decode(&vaa).map_err(|_| Error::InvalidHex)?;
+            let vaa = serde_wormhole::from_slice_with_payload::<wormhole::Vaa<()>>(&vaa);
+            let vaa = vaa.map_err(|_| Error::InvalidVaa)?;
+            let (vaa, _rest) = vaa;
+
+            // Convert to local VAA type to catch APi changes.
+            let vaa = Vaa::from(vaa);
+
+            // Prevent VAA re-execution.
+            if self.executed_gov_sequences.contains(&vaa.sequence) {
+                return Err(Error::VaaVerificationFailed);
+            }
+
+            // Confirm the VAA is coming from a trusted source chain.
+            if self.gov_source
+                != (Source {
+                    emitter:            vaa.emitter_address,
+                    pyth_emitter_chain: vaa.emitter_chain,
+                })
+            {
+                return Err(Error::UnknownSource);
+            }
+
+            // Insert before calling Wormhole to prevent re-execution. If we wait until after the
+            // Wormhole call we could end up with multiple VAA's with the same sequence being
+            // executed in parallel.
+            self.executed_gov_sequences.insert(&vaa.sequence);
+        }
+
+        // Verify VAA and refund the caller in case of failure.
+        ext_wormhole::ext(self.wormhole.clone())
+            .with_static_gas(Gas(30_000_000_000_000))
```

Comment:
how do we know that this quantity is enough? does wormhole guarantee this?

what happens if there isn't enough gas?

Reply:
also extract out the gas number to a constant variable? seems like this number is used in a bunch of places